server/src/main/java/org/elasticsearch/index/engine/TranslogDirectoryReader.java

```diff
@@ -0,0 +1,540 @@
+                    }
+                }
+                if (sourceBytes != null) {
+                    visitor.binaryField(FAKE_SOURCE_FIELD, BytesReference.toBytes(sourceBytes));
+                }
+            }
+            if (operation.routing() != null && visitor.needsField(FAKE_ROUTING_FIELD) == StoredFieldVisitor.Status.YES) {
+                visitor.stringField(FAKE_ROUTING_FIELD, operation.routing().getBytes(StandardCharsets.UTF_8));
+            }
+            if (visitor.needsField(FAKE_ID_FIELD) == StoredFieldVisitor.Status.YES) {
+                final byte[] id = new byte[uid.length];
+                System.arraycopy(uid.bytes, uid.offset, id, 0, uid.length);
+                visitor.binaryField(FAKE_ID_FIELD, id);
+            }
+        }
+
+        @Override
+        protected synchronized void doClose() throws IOException {
+            IOUtils.close(delegate.get(), directory);
+        }
+    }
+
+    private static class FakeTerms extends Terms {
+        private final BytesRef uid;
+
+        FakeTerms(BytesRef uid) {
+            this.uid = uid;
+        }
+
+        @Override
+        public TermsEnum iterator() throws IOException {
+            return new FakeTermsEnum(uid);
+        }
+
+        @Override
+        public long size() throws IOException {
+            return 1;
+        }
+
+        @Override
+        public long getSumTotalTermFreq() throws IOException {
+            return 1;
+        }
+
+        @Override
+        public long getSumDocFreq() throws IOException {
+            return 1;
+        }
+
+        @Override
+        public int getDocCount() throws IOException {
+            return 1;
+        }
+
+        @Override
+        public boolean hasFreqs() {
+            return false;
+        }
+
+        @Override
+        public boolean hasOffsets() {
+            return false;
+        }
+
+        @Override
+        public boolean hasPositions() {
+            return false;
+        }
+
+        @Override
+        public boolean hasPayloads() {
+            return false;
+        }
+    }
+
+    private static class FakeTermsEnum extends BaseTermsEnum {
+        private final BytesRef term;
+        private long position = -1;
+
+        FakeTermsEnum(BytesRef term) {
+            this.term = term;
+        }
+
+        @Override
+        public SeekStatus seekCeil(BytesRef text) throws IOException {
+            int cmp = text.compareTo(term);
+            if (cmp == 0) {
+                position = 0;
+                return SeekStatus.FOUND;
+            } else if (cmp < 0) {
+                position = 0;
+                return SeekStatus.NOT_FOUND;
+            }
+            position = Long.MAX_VALUE;
+            return SeekStatus.END;
+        }
+
+        @Override
+        public void seekExact(long ord) throws IOException {
+            position = ord;
+        }
+
+        @Override
+        public BytesRef term() throws IOException {
+            return term;
```

Comment:
I find it slightly odd that `seekCeil` looks like it actually seeks, whereas here we do not take `position` into account. I am sure it works, but I would prefer to either just assert that `text` is the same as `term` in `seekCeil` or take position into account in the methods here - just for consistency.

Reply:
Thinking more, this looks good as is. We could add `assert position == 0`, that might have lead me on the right track when reading this.